Mathlib/SetTheory/Ordinal/Nimber.lean

```diff
@@ -357,4 +361,210 @@ theorem add_trichotomy {a b c : Nimber} (h : a + b + c ≠ 0) :
   · rw [← hx'] at hx
     exact Or.inr <| Or.inr hx
 
+theorem two_nsmul (a : Nimber) : 2 • a = 0 := by
+  rw [succ_nsmul, one_nsmul, add_self]
+
+theorem two_zsmul (a : Nimber) : (2 : ℤ) • a = 0 :=
+  two_nsmul a
+
+theorem add_eq_iff_eq_add : a + b = c ↔ a = c + b :=
+  sub_eq_iff_eq_add
+
+theorem eq_add_iff_add_eq : a = b + c ↔ a + c = b :=
+  eq_sub_iff_add_eq
+
+/-! ### Nimber multiplication -/
+
+/-- Nimber multiplication is recursively defined so that `a * b` is the smallest nimber not equal to
+`a' * b + a * b' + a' * b'` for `a' < a` and `b' < b`. -/
+-- We write the binders like this so that the termination checker works.
+protected def mul (a b : Nimber.{u}) : Nimber.{u} :=
+  sInf {x | ∃ a', ∃ (_ : a' < a), ∃ b', ∃ (_ : b' < b),
+    Nimber.mul a' b + Nimber.mul a b' + Nimber.mul a' b' = x}ᶜ
+termination_by (a, b)
+
+instance : Mul Nimber :=
+  ⟨Nimber.mul⟩
+
+theorem mul_def (a b : Nimber) :
+    a * b = sInf {x | ∃ a' < a, ∃ b' < b, a' * b + a * b' + a' * b' = x}ᶜ := by
+  change Nimber.mul a b = _
+  rw [Nimber.mul]
+  simp_rw [exists_prop]
+  rfl
+
+/-- The set in the definition of `Nimber.mul` is nonempty. -/
+private theorem mul_nonempty (a b : Nimber.{u}) :
+    {x | ∃ a' < a, ∃ b' < b, a' * b + a * b' + a' * b' = x}ᶜ.Nonempty := by
+  convert compl_nonempty_of_small
+    ((fun x ↦ x.1 * b + a * x.2 + x.1 * x.2) '' Set.Iio a ×ˢ Set.Iio b)
+  ext
+  simp_rw [Set.mem_setOf_eq, Set.mem_image, Set.mem_prod, Set.mem_Iio, Prod.exists]
+  tauto
+
+theorem exists_of_lt_mul (h : c < a * b) : ∃ a' < a, ∃ b' < b, a' * b + a * b' + a' * b' = c := by
+  rw [mul_def] at h
+  have := not_mem_of_lt_csInf h ⟨_, bot_mem_lowerBounds _⟩
+  rwa [Set.not_mem_compl_iff] at this
+
+theorem mul_le_of_forall_ne (h : ∀ a' < a, ∀ b' < b, a' * b + a * b' + a' * b' ≠ c) :
+    a * b ≤ c := by
+  by_contra! h'
+  have := exists_of_lt_mul h'
+  tauto
+
+instance : MulZeroClass Nimber where
+  mul_zero a := by
+    rw [← Nimber.le_zero]
+    apply mul_le_of_forall_ne
+    intro _ _ _ h
+    exact (Nimber.not_lt_zero _ h).elim
+  zero_mul a := by
+    rw [← Nimber.le_zero]
+    apply mul_le_of_forall_ne
+    intro _ h
+    exact (Nimber.not_lt_zero _ h).elim
+
+private theorem mul_ne_of_lt : ∀ a' < a, ∀ b' < b, a' * b + a * b' + a' * b' ≠ a * b := by
+  have H := csInf_mem (mul_nonempty a b)
+  rw [← mul_def] at H
+  simpa using H
+
+instance : NoZeroDivisors Nimber := by
+  constructor
+  intro a b h
+  by_contra! hab
+  iterate 2 rw [← Nimber.pos_iff_ne_zero] at hab
+  apply (mul_ne_of_lt _ hab.1 _ hab.2).symm
+  simpa only [zero_add, mul_zero, zero_mul]
+
+protected theorem mul_comm (a b : Nimber) : a * b = b * a := by
+  apply le_antisymm <;>
```

Comment:
Instead of using `<;>`, can you use a `wlog` or a `suffices` for the inequality?

That way you produce one proof and use it twice, rather than producing the same proof twice.

Reply:
I can't prove `a * b ≤ b * a` without using the full induction hypothesis `a * b = b * a` for smaller values. In fact, you'll notice the proofs of both directions are slightly distinct; I have to rewrite backwards on goal 2 to ensure I'm using `x * y = y * x` for `x < a` and `y < b`, instead of `y * x = x * y`.